rmarkdown/parameters_csv.md
Using parameters from csv
================
Amy Heather
2025-07-11

## Set-up

Install the latest version of the local simulation package. If running
sequentially, `devtools::load_all()` is sufficient. If running in
parallel, you must use `devtools::install()` and then `library(simulation)`.

``` r
devtools::load_all()
```

    ## ℹ Loading simulation

``` r
start_time <- Sys.time()
```

## Creating parameter class

We can set up classes based on CSV, replacing the default inputs with
those from the csv.

If you were only planning to use parameters from CSV, then you could
remove the default inputs from the classes in `parameters.R` altogether.

We use the parameter function to process values from the CSV, ensuring
that all required parameter names are present and that no unexpected
arguments are included. This approach leverages the function’s built-in
validation to catch any issues with argument names or types. If the
parameter classes themselves have validation, those checks will also be
performed when the function is called.

``` r
#' Use parameter function to create parameter list using values from dataframe.
#'
#' @param data Dataframe with columns "unit", "parameter", "type", "mean" and
#' "sd".
#' @param unit Unit name to filter by ("asu" or "rehab").
#' @param parameter Parameter name to filter by ("iat", "los" or "routing").
#' @param param_function Function to run
#'
#' @return Named list of parameters

init_param_class <- function(data, unit, parameter, param_function) {

  # Filter data to specified unit and parameter
  df_subset <- data[data[["unit"]] == unit &
                      data[["parameter"]] == parameter, ]

  # Create named list of parameter values.
  # If all SD values are missing, use only means and name by 'type'.
  # Otherwise, include both mean and sd for each type, with names like
  # 'type_mean' and 'type_sd'.
  if (all(is.na(df_subset$sd))) {
    param_list <- as.list(setNames(df_subset$mean, df_subset$type))
  } else {
    param_list <- list()
    for (i in seq_len(nrow(df_subset))) {
      df_row <- df_subset[i, ]
      param_list[[paste0(df_row$type, "_mean")]] <- df_row$mean
      param_list[[paste0(df_row$type, "_sd")]] <- df_row$sd
    }
  }

  # Run parameter function using list
  do.call(param_function, param_list)
}
```

``` r
#' Generate named_list with create_parameters() using values loaded from a CSV
#' file.
#'
#' @param csv_path Path to csv file containing the parameters. Should have
#' columns "unit", "parameter", "type", "mean" and "sd". Missing values should
#' be marked as "NA".
#'
#' @return Named list generated by create_parameters()

setup_param_from_csv <- function(csv_path) {
  # Load parameter data from CSV, treating "NA" as missing values
  param_data <- read.csv(csv_path, na.strings = "NA")

  # Specify mappings of create_parameters() arguments to their corresponding
  # units, parameter types and parameter classes
  param_specs <- list(
    list(name = "asu_arrivals",
         unit = "asu",
         parameter = "iat",
         param_function = create_asu_arrivals),
    list(name = "rehab_arrivals",
         unit = "rehab",
         parameter = "iat",
         param_function = create_rehab_arrivals),
    list(name = "asu_los",
         unit = "asu",
         parameter = "los",
         param_function = create_asu_los),
    list(name = "rehab_los",
         unit = "rehab",
         parameter = "los",
         param_function = create_rehab_los),
    list(name = "asu_routing",
         unit = "asu",
         parameter = "routing",
         param_function = create_asu_routing),
    list(name = "rehab_routing",
         unit = "rehab",
         parameter = "routing",
         param_function = create_rehab_routing)
  )
  param_kwargs <- list()
  for (spec in param_specs) {
    param_kwargs[[spec$name]] <- init_param_class(
      data = param_data, unit = spec$unit, parameter = spec$parameter,
      param_function = spec$param_function
    )
  }

  do.call(create_parameters, param_kwargs)
}
```

``` r
setup_param_from_csv(file.path("..", "inputs", "parameters.csv"))
```

    ## $asu_arrivals
    ## $asu_arrivals$stroke
    ## [1] 1.2
    ## 
    ## $asu_arrivals$tia
    ## [1] 9.3
    ## 
    ## $asu_arrivals$neuro
    ## [1] 3.6
    ## 
    ## $asu_arrivals$other
    ## [1] 3.2
    ## 
    ## 
    ## $rehab_arrivals
    ## $rehab_arrivals$stroke
    ## [1] 21.8
    ## 
    ## $rehab_arrivals$neuro
    ## [1] 31.7
    ## 
    ## $rehab_arrivals$other
    ## [1] 28.6
    ## 
    ## 
    ## $asu_los
    ## $asu_los$stroke_no_esd
    ## $asu_los$stroke_no_esd$mean
    ## [1] 7.4
    ## 
    ## $asu_los$stroke_no_esd$sd
    ## [1] "8.61"
    ## 
    ## 
    ## $asu_los$stroke_esd
    ## $asu_los$stroke_esd$mean
    ## [1] 4.6
    ## 
    ## $asu_los$stroke_esd$sd
    ## [1] "4.8"
    ## 
    ## 
    ## $asu_los$stroke_mortality
    ## $asu_los$stroke_mortality$mean
    ## [1] 7
    ## 
    ## $asu_los$stroke_mortality$sd
    ## [1] 8.7
    ## 
    ## 
    ## $asu_los$tia
    ## $asu_los$tia$mean
    ## [1] 1.8
    ## 
    ## $asu_los$tia$sd
    ## [1] "2.3"
    ## 
    ## 
    ## $asu_los$neuro
    ## $asu_los$neuro$mean
    ## [1] 4
    ## 
    ## $asu_los$neuro$sd
    ## [1] "5"
    ## 
    ## 
    ## $asu_los$other
    ## $asu_los$other$mean
    ## [1] 3.8
    ## 
    ## $asu_los$other$sd
    ## [1] "5.2"
    ## 
    ## 
    ## 
    ## $rehab_los
    ## $rehab_los$stroke_no_esd
    ## $rehab_los$stroke_no_esd$mean
    ## [1] 28.4
    ## 
    ## $rehab_los$stroke_no_esd$sd
    ## [1] "27.2"
    ## 
    ## 
    ## $rehab_los$stroke_esd
    ## $rehab_los$stroke_esd$mean
    ## [1] 30.3
    ## 
    ## $rehab_los$stroke_esd$sd
    ## [1] "2un3.1"
    ## 
    ## 
    ## $rehab_los$tia
    ## $rehab_los$tia$mean
    ## [1] 18.7
    ## 
    ## $rehab_los$tia$sd
    ## [1] "23.5"
    ## 
    ## 
    ## $rehab_los$neuro
    ## $rehab_los$neuro$mean
    ## [1] 27.6
    ## 
    ## $rehab_los$neuro$sd
    ## [1] "28.4"
    ## 
    ## 
    ## $rehab_los$other
    ## $rehab_los$other$mean
    ## [1] 16.1
    ## 
    ## $rehab_los$other$sd
    ## [1] "14.1"
    ## 
    ## 
    ## 
    ## $asu_routing
    ## $asu_routing$stroke
    ## $asu_routing$stroke$rehab
    ## [1] 0.24
    ## 
    ## $asu_routing$stroke$esd
    ## [1] 0.13
    ## 
    ## $asu_routing$stroke$other
    ## [1] 0.63
    ## 
    ## 
    ## $asu_routing$tia
    ## $asu_routing$tia$rehab
    ## [1] 0.01
    ## 
    ## $asu_routing$tia$esd
    ## [1] 0.01
    ## 
    ## $asu_routing$tia$other
    ## [1] 0.98
    ## 
    ## 
    ## $asu_routing$neuro
    ## $asu_routing$neuro$rehab
    ## [1] 0.11
    ## 
    ## $asu_routing$neuro$esd
    ## [1] 0.05
    ## 
    ## $asu_routing$neuro$other
    ## [1] 0.84
    ## 
    ## 
    ## $asu_routing$other
    ## $asu_routing$other$rehab
    ## [1] 0.05
    ## 
    ## $asu_routing$other$esd
    ## [1] 0.1
    ## 
    ## $asu_routing$other$other
    ## [1] 0.85
    ## 
    ## 
    ## 
    ## $rehab_routing
    ## $rehab_routing$stroke
    ## $rehab_routing$stroke$esd
    ## [1] 0.4
    ## 
    ## $rehab_routing$stroke$other
    ## [1] 0.6
    ## 
    ## 
    ## $rehab_routing$tia
    ## $rehab_routing$tia$esd
    ## [1] 0
    ## 
    ## $rehab_routing$tia$other
    ## [1] 1
    ## 
    ## 
    ## $rehab_routing$neuro
    ## $rehab_routing$neuro$esd
    ## [1] 0.09
    ## 
    ## $rehab_routing$neuro$other
    ## [1] 0.91
    ## 
    ## 
    ## $rehab_routing$other
    ## $rehab_routing$other$esd
    ## [1] 0.13
    ## 
    ## $rehab_routing$other$other
    ## [1] 0.88
    ## 
    ## 
    ## 
    ## $warm_up_period
    ## [1] 1095
    ## 
    ## $data_collection_period
    ## [1] 1825
    ## 
    ## $number_of_runs
    ## [1] 150
    ## 
    ## $scenario_name
    ## NULL
    ## 
    ## $cores
    ## [1] 1
    ## 
    ## $log_to_console
    ## [1] FALSE
    ## 
    ## $log_to_file
    ## [1] FALSE
    ## 
    ## $file_path
    ## NULL

## Calculate run time

``` r
# Get run time in seconds
end_time <- Sys.time()
runtime <- as.numeric(end_time - start_time, units = "secs")

# Display converted to minutes and seconds
minutes <- as.integer(runtime / 60L)
seconds <- as.integer(runtime %% 60L)
cat(sprintf("Notebook run time: %dm %ds", minutes, seconds))
```

    ## Notebook run time: 0m 0s


NEWS.md

# Stroke capacity planning model: R DES RAP 0.1.0

🌱 First release of the R stroke model.

## New features

* Implementation of the stroke model in R as a package (`R/`).
* Reproduction of paper results (`rmarkdown/analysis.Rmd`).
* Other RMarkdown files to generate results for tests, demonstrate logging, and how parameters can be loaded from a csv (`rmarkdown/`).
* Back, functional and unit tests (`tests/`).
* Other features include continuous integration (tests and linting), checklists, R environment, and other metadata.

R/parameters.R
#' Acute Stroke Unit (ASU) arrival intervals (days).
#'
#' For example, a value of 1.2 means a new admission every 1.2 days.
#'
#' @param stroke Numeric. Mean days between stroke patient arrivals.
#' @param tia Numeric. Mean days between transient ischaemic attack (TIA)
#' patient arrivals.
#' @param neuro Numeric. Mean days between complex neurological patient
#' arrivals.
#' @param other Numeric. Mean days between other patient arrivals.
#'
#' @return A named list of arrival intervals for ASU.
#' @export

create_asu_arrivals <- function(
  stroke = 1.2, tia = 9.3, neuro = 3.6, other = 3.2
) {
  return(as.list(environment()))
}

#' Rehabilitation unit arrival intervals (days).
#'
#' For example, a value of 21.8 means a new admission every 21.8 days.
#'
#' @param stroke Numeric. Mean days between stroke patient arrivals.
#' @param neuro Numeric. Mean days between complex neurological patient
#' arrivals.
#' @param other Numeric. Mean days between other patient arrivals.
#'
#' @return A named list of arrival intervals for rehabilitation unit.
#' @export

create_rehab_arrivals <- function(
  stroke = 21.8, neuro = 31.7, other = 28.6
) {
  return(as.list(environment()))
}

#' Acute Stroke Unit (ASU) length of stay (LOS) distributions (days).
#'
#' Mean and standard deviation (SD) of LOS in days in the ASU.
#'
#' @param stroke_noesd_mean Numeric. Mean LOS for stroke patients without early
#' supported discharged (ESD).
#' @param stroke_noesd_sd Numeric. SD LOS for stroke patients without ESD.
#' @param stroke_esd_mean Numeric. Mean LOS for stroke patients with ESD.
#' @param stroke_esd_sd Numeric. SD LOS for stroke patients with ESD.
#' @param stroke_mortality_mean Numeric. Mean LOS for stroke patients who pass
#' away.
#' @param stroke_mortality_sd Numeric. SD LOS for stroke patients who pass
#' away.
#' @param tia_mean Numeric. Mean LOS for transient ischemic attack (TIA)
#' patients.
#' @param tia_sd Numeric. SD LOS for TIA patients.
#' @param neuro_mean Numeric. Mean LOS for complex neurological patients
#' @param neuro_sd Numeric. SD LOS for complex neurological patients
#' @param other_mean Numeric. Mean LOS for other patients.
#' @param other_sd Numeric. SD LOS for other patients.
#'
#' @return A named list of LOS distributions for ASU.
#' @export

create_asu_los <- function(
  stroke_noesd_mean = 7.4, stroke_noesd_sd = 8.61,
  stroke_esd_mean = 4.6, stroke_esd_sd = 4.8,
  stroke_mortality_mean = 7.0, stroke_mortality_sd = 8.7,
  tia_mean = 1.8, tia_sd = 2.3,
  neuro_mean = 4.0, neuro_sd = 5.0,
  other_mean = 3.8, other_sd = 5.2
) {
  list(
    stroke_noesd = list(mean = stroke_noesd_mean, sd = stroke_noesd_sd),
    stroke_esd = list(mean = stroke_esd_mean, sd = stroke_esd_sd),
    stroke_mortality = list(mean = stroke_mortality_mean,
                            sd = stroke_mortality_sd),
    tia = list(mean = tia_mean, sd = tia_sd),
    neuro = list(mean = neuro_mean, sd = neuro_sd),
    other = list(mean = other_mean, sd = other_sd)
  )
}

#' Rehabilitation unit length of stay (LOS) distributions (days).
#'
#' @param stroke_noesd_mean Numeric. Mean LOS for stroke patients without early
#' supported discharged (ESD).
#' @param stroke_noesd_sd Numeric. SD LOS for stroke patients without ESD.
#' @param stroke_esd_mean Numeric. Mean LOS for stroke patients with ESD.
#' @param stroke_esd_sd Numeric. SD LOS for stroke patients with ESD.
#' @param tia_mean Numeric. Mean LOS for transient ischemic attack (TIA)
#' patients.
#' @param tia_sd Numeric. SD LOS for TIA patients.
#' @param neuro_mean Numeric. Mean LOS for complex neurological patients
#' @param neuro_sd Numeric. SD LOS for complex neurological patients
#' @param other_mean Numeric. Mean LOS for other patients.
#' @param other_sd Numeric. SD LOS for other patients.
#'
#' @return A named list of LOS distributions for rehabilitation unit.
#' @export

create_rehab_los <- function(
  stroke_noesd_mean = 28.4, stroke_noesd_sd = 27.2,
  stroke_esd_mean = 30.3, stroke_esd_sd = 23.1,
  tia_mean = 18.7, tia_sd = 23.5,
  neuro_mean = 27.6, neuro_sd = 28.4,
  other_mean = 16.1, other_sd = 14.1
) {
  list(
    stroke_noesd = list(mean = stroke_noesd_mean, sd = stroke_noesd_sd),
    stroke_esd = list(mean = stroke_esd_mean, sd = stroke_esd_sd),
    tia = list(mean = tia_mean, sd = tia_sd),
    neuro = list(mean = neuro_mean, sd = neuro_sd),
    other = list(mean = other_mean, sd = other_sd)
  )
}

#' ASU routing probabilities.
#'
#' Probabilities of each patient type being transferred from the acute
#' stroke unit (ASU) to other destinations.
#'
#' @param stroke_rehab Numeric. Probability stroke patient to rehab.
#' @param stroke_esd Numeric. Probability stroke patient to early supported
#' discharge (ESD) services.
#' @param stroke_other Numeric. Probability stroke patient to other
#' destinations (e.g., own home, care home, mortality).
#' @param tia_rehab Numeric. Probability transient ischemic attack (TIA)
#' patient to rehab.
#' @param tia_esd Numeric. Probability TIA patient to ESD.
#' @param tia_other Numeric. Probability TIA patient to other.
#' @param neuro_rehab Numeric. Probability complex neurological patient to
#' rehab.
#' @param neuro_esd Numeric. Probability complex neurological patient to ESD.
#' @param neuro_other Numeric. Probability complex neurological patient to
#' other.
#' @param other_rehab Numeric. Probability other patient to rehab.
#' @param other_esd Numeric. Probability other patient to ESD.
#' @param other_other Numeric. Probability other patient to other.
#'
#' @return A named list of routing probabilities for ASU.
#' @export

create_asu_routing <- function(
  stroke_rehab = 0.24, stroke_esd = 0.13, stroke_other = 0.63,
  tia_rehab = 0.01, tia_esd = 0.01, tia_other = 0.98,
  neuro_rehab = 0.11, neuro_esd = 0.05, neuro_other = 0.84,
  other_rehab = 0.05, other_esd = 0.10, other_other = 0.85
) {
  list(
    stroke = list(rehab = stroke_rehab, esd = stroke_esd, other = stroke_other),
    tia = list(rehab = tia_rehab, esd = tia_esd, other = tia_other),
    neuro = list(rehab = neuro_rehab, esd = neuro_esd, other = neuro_other),
    other = list(rehab = other_rehab, esd = other_esd, other = other_other)
  )
}

#' Rehabilitation unit routing probabilities.
#'
#' Probabilities of each patient type being transferred from the rehabilitation
#' unit to other destinations.
#'
#' @param stroke_esd Numeric. Probability stroke patient to early supported
#' discharge (ESD) services.
#' @param stroke_other Numeric. Probability stroke patient to other
#' destinations (e.g., own home, care home, mortality).
#' @param tia_esd Numeric. Probability transient ischemic attack (TIA) patient
#' to ESD.
#' @param tia_other Numeric. Probability TIA patient to other.
#' @param neuro_esd Numeric. Probability complex neurological patient to ESD.
#' @param neuro_other Numeric. Probability complex neurological patient to
#' other.
#' @param other_esd Numeric. Probability other patient to ESD.
#' @param other_other Numeric. Probability other patient to other.
#'
#' @return A named list of routing probabilities for rehabilitation unit.
#' @export

create_rehab_routing <- function(
  stroke_esd = 0.40, stroke_other = 0.60,
  tia_esd = 0L, tia_other = 1L,
  neuro_esd = 0.09, neuro_other = 0.91,
  other_esd = 0.13, other_other = 0.88
) {
  list(
    stroke = list(esd = stroke_esd, other = stroke_other),
    tia = list(esd = tia_esd, other = tia_other),
    neuro = list(esd = neuro_esd, other = neuro_other),
    other = list(esd = other_esd, other = other_other)
  )
}

#' Generate complete parameter list for simulation.
#'
#' @param asu_arrivals List. Acute stroke unit (ASU) arrival intervals.
#' @param rehab_arrivals List. Rehabilitation unit arrival intervals.
#' @param asu_los List. ASU length of stay (LOS) distributions.
#' @param rehab_los List. Rehabilitation unit LOS distributions.
#' @param asu_routing List. ASU routing probabilities.
#' @param rehab_routing List. Rehabilitation unit routing probabilities.
#' @param warm_up_period Integer. Length of warm-up period (days).
#' @param data_collection_period Integer. Length of data collection period
#' (days).
#' @param number_of_runs Integer. Number of simulation runs.
#' @param scenario_name Label for scenario (int|float|string).
#' @param cores Integer. Number of CPU cores to use.
#' @param log_to_console Whether to print activity log to console.
#' @param log_to_file Whether to save activity log to file.
#' @param file_path Path to save log to file.
#'
#' @return A named list of all simulation parameters.
#' @export

create_parameters <- function(
  asu_arrivals = create_asu_arrivals(),
  rehab_arrivals = create_rehab_arrivals(),
  asu_los = create_asu_los(),
  rehab_los = create_rehab_los(),
  asu_routing = create_asu_routing(),
  rehab_routing = create_rehab_routing(),
  warm_up_period = 365L * 3L,  # 3 years
  data_collection_period = 365L * 5L,  # 5 years
  number_of_runs = 150L,
  scenario_name = NULL,
  cores = 1L,
  log_to_console = FALSE,
  log_to_file = FALSE,
  file_path = NULL
) {
  return(as.list(environment()))
}
